pyo3-macros-backend/src/pyenum.rs

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2017-present PyO3 Project and Contributors
+
+use crate::common::impl_extractext;
+use crate::pyimpl::{PyClassImplBuilder, PyClassMethodsType};
+use proc_macro2::TokenStream;
+use quote::quote;
+
+pub fn build_py_enum(
+    enum_: &syn::ItemEnum,
+    methods_type: PyClassMethodsType,
+) -> syn::Result<TokenStream> {
+    let mut variants = Vec::new();
+
+    for variant in enum_.variants.iter() {
+        if !variant.fields.is_empty() {
+            return Err(syn::Error::new_spanned(
+                variant,
+                "#[pyenum] only supports unit enums",
+            ));
+        }
+        if let Some((_, syn::Expr::Lit(lit))) = &variant.discriminant {
+            variants.push((variant.ident.clone(), lit.clone()))
+        } else {
+            return Err(syn::Error::new_spanned(
+                variant,
+                "#[pyenum] requires explicit discriminant (MyVal = 4)",
+            ));
+        }
+    }
+
+    impl_enum(&enum_.ident, variants, methods_type)
+}
+
+fn impl_enum(
+    enum_: &syn::Ident,
+    variants: Vec<(syn::Ident, syn::ExprLit)>,
+    methods_type: PyClassMethodsType,
+) -> syn::Result<TokenStream> {
+    let enum_cls = impl_class(enum_, None, methods_type)?;
+    let variant_names: Vec<syn::Ident> = variants
+        .iter()
+        .map(|(ident, _)| variant_enumname(enum_, ident))
+        .collect::<syn::Result<Vec<_>>>()?;
+
+    let variant_cls = variants
+        .iter()
+        .map(|(ident, _)| impl_class(ident, Some(enum_), methods_type))
+        .collect::<syn::Result<Vec<_>>>()?;
+    let variant_consts = variants
+        .iter()
+        .map(|(ident, _)| impl_const(enum_, ident))
+        .collect::<syn::Result<Vec<_>>>()?;
+
+    let to_py = impl_to_py(enum_, &variants)?;
+
+    Ok(quote! {
+
+        #enum_cls
+
+        #(
+            struct #variant_names;
+        )*
+
+        #(
+            #variant_cls
+        )*
+
+        #to_py
+
+        #[pymethods]
+        impl #enum_ {
+            #(
+                #[allow(non_upper_case_globals)]
+                #variant_consts
+            )*
+        }
```

Comment:
This is clever, but I think the downside of this is that without `multiple-pymethods` feature it won't be possible for users to add their own `pymethods` block.

You should instead use [`gen_py_const`](https://github.com/PyO3/pyo3/blob/2f6ea2f96cf8d9f213ea6c8b047e95d5e4aecaca/pyo3-macros-backend/src/pyimpl.rs#L121) to create "method definitions" which then need to be fed into the collection machinery via a "methods trait" - probably `PyClassDescriptors` can be "abused" here (that's the one which collects all the `#[pyo3(get, set)]` annotations from regular `#[pyclass]`).

Let me know if you want me to give some more pointers - this code is rather messy and not well documented as it's in a permanent state of evolution / hackery :)

Reply:
Let me try to figure that out. Meanwhile there's a new problem: I need to implement __eq__ (via __richcmp__) for the enum to be usable, but I'll need a way to do that.

Reply:
Mmm ultimately you'll need to generate the code to create a `PyType_Slot` struct which points to the implementation of `__richcmp__` you want, which then gets fed into the collection machinery as part of the `for_each_proto_slot` function.

There's some code in `pyo3-macros-backend/src/pymethod.rs` to which handles the magic methods as part of `pymethods` which might be useful if you need to generate a `__richcmp__` method. Or alternatively maybe you're able to hard-code an "enum protocols trait" in `src/class/impl_.rs` which you can then add to enum classes only using the `PyClassImplBuilder`.

Hopefully again those are some notes to get started reading the right code - please do ask more questions as you need!

Reply:
I have an idea, how about we derive from the enum and provide `__richcmp__` in the subclass? That's less complex, and we don't have to handle the situation where the user want to provide their own `__richcmp__`.
Equivalent code in Python:
```Python
class Enum:
    @classmethod
    @property
    def Var1(cls):
        return EnumVar1()
...

class EnumVar1(Enum):
    def __eq__(self, other):
        # Return true only if the other is a `Enum` and represents Var1
```

Reply:
Perhaps, I think if the user provides their own `__richcmp__` and it overrides our default one that's fine. Let's see how this goes.

Reply:
Another thought could be to enforce singletons for each variant, this removing need for `__richcmp__`

Reply:
I wanted to use subclass approach because I didn't understand how PyO3 collects methods and thought if we provided `__richcmp__`, the user couldn't create their own. Now I've read the implementation, I know it's not the case and I think I like introducing a new trait. 
I don't know how to enforce singletons.